src/v/cluster/tests/partition_allocator_tests.cc

```diff
@@ -85,6 +85,41 @@ FIXTURE_TEST(unregister_node, partition_allocator_fixture) {
     BOOST_REQUIRE_EQUAL(allocator.state().available_nodes(), 2);
 }
 
+FIXTURE_TEST(allocation_over_core_capacity, partition_allocator_fixture) {
+    const auto partition_count
+      = partition_allocator_fixture::partitions_per_shard + 1;
+    register_node(0, 1);
+    auto result
+      = allocator.allocate(make_allocation_request(partition_count, 1)).get();
+    BOOST_REQUIRE(result.has_error());
+    BOOST_REQUIRE_EQUAL(
+      result.assume_error(),
+      cluster::make_error_code(
+        cluster::errc::topic_invalid_partitions_core_limit));
```

Comment:
nitpick/FYI: a `std::error_code` has an implicit constructor that that is enabled when `is_error_code_enum<T>`, which is true for `cluster::errc`. That constructor calls `make_error_code` using ADL. So the following code is perfectly acceptable, as the comparison will construct a `std::error_code`:
```suggestion
      result.assume_error(), cluster::errc::topic_invalid_partitions_fd_limit);
```